quicklens/sims/util.py

```diff
@@ -36,3 +38,218 @@ def hashfail(msg=None):
+
+        if self.docov == True:
+            self.bcl_cv_sum += np.outer(tml, tml)
+
+    def avg(self):
+        return spec.bcl(self.lbins, {'cl': self.bcl_av_sum / self.nsims})
+
+    def std(self):
+        return spec.bcl(self.lbins, {'cl': np.sqrt((self.bcl_sq_sum - self.bcl_av_sum**2 / self.nsims) / self.nsims)})
+
+    def var(self):
+        return spec.bcl(self.lbins, {'cl': (self.bcl_sq_sum - self.bcl_av_sum**2 / self.nsims) / self.nsims})
+
+    def cov(self):
+        assert(self.docov == True)
+        return (self.bcl_cv_sum - np.outer(self.bcl_av_sum, self.bcl_av_sum) / self.nsims) / self.nsims
+
+    def save(self, fname):
+        '''
+        Turn this object into a dictionary so it can be Pickled.
+        Note: lambda functions can't be pickled, so you have to pass in the same lambda function on read-in.
+        '''
+        ml_stats_dict = {'lbins' : self.lbins,
+                         'wb' : self.wb,
+                         'docov' : self.docov,
+                         'nsims' : self.nsims,
+                         'bcl_av_sum' : self.bcl_av_sum,
+                         'bcl_sq_sum' : self.bcl_sq_sum}
+        if hasattr(self, 'bcl_cv_sum'): ml_stats_dict['bcl_cv_sum'] = self.bcl_cv_sum
+
+        if not os.path.exists(fname):
+            pk.dump(ml_stats_dict, open(fname,'w'))
+        else:
+            raise IOError("File already exists!")
+
+    def load(self, fname):
+        '''
+        Load an object from a saved pickle file.
+        Note: you must pass in the same lambda function for w as was used for the original object
+        '''
+        ml_stats_dict    = pk.load(open(fname,'r'))
+        self.lbins       = ml_stats_dict['lbins']
+        self.wb          = ml_stats_dict['wb']
+        self.docov       = ml_stats_dict['docov']
+        self.nsims       = ml_stats_dict['nsims']
+        self.bcl_av_sum  = ml_stats_dict['bcl_av_sum']
+        self.bcl_sq_sum  = ml_stats_dict['bcl_sq_sum']
+        if ('bcl_cv_sum' in ml_stats_dict.keys()): self.bcl_cv_sum = ml_stats_dict['bcl_cv_sum']
+
+    def plot_fill_between(self, t=lambda l: 1.0, m=None, **kwargs):
+        '''
+        plot this spectrum as a histogram
+        t,            l-dependent coefficient.  e.g., t = lambda l : (l*(l+1.))**2 / (2.*np.pi) * 1.e7
+        m,            the spectrum to be plotted
+        '''
+        ls = 0.5 * (self.lbins[0:-1] + self.lbins[1:])  # bin centers
+        if m == None:
+            m = self.avg().cl
+
+        plt.fill_between(ls, t(ls) * (m - self.std().cl), t(ls) * (m + self.std().cl), **kwargs)
+
+    def plot_error_bars(self, t=lambda l: 1.0, m=None, p=plt.errorbar, **kwargs):
+        ls = 0.5 * (self.lbins[0:-1] + self.lbins[1:])  # bin centers
+        if m == None:
+            m = self.avg().cl
+
+        p(ls, t(ls) * m, yerr=(t(ls) * self.std().cl), **kwargs)
+
+class ml_cross_stats():
+    '''
+    Class to gater stats from a cross-spectrum.
+    This emulates sims.util.ml_stats()
+    '''
```

Comment:
A lot of the code here is identical to that in ml_stats -- WDYT about using it as a base class?